src/compiler/parser.ts

```diff
@@ -6079,14 +6110,29 @@ namespace ts {
         }
 
         function setExternalModuleIndicator(sourceFile: SourceFile) {
-            sourceFile.externalModuleIndicator = forEach(sourceFile.statements, node =>
-                hasModifier(node, ModifierFlags.Export)
-                    || node.kind === SyntaxKind.ImportEqualsDeclaration && (<ImportEqualsDeclaration>node).moduleReference.kind === SyntaxKind.ExternalModuleReference
-                    || node.kind === SyntaxKind.ImportDeclaration
-                    || node.kind === SyntaxKind.ExportAssignment
-                    || node.kind === SyntaxKind.ExportDeclaration
+            // Usually we'd like to avoid a full tree walk, but it's possible
+            // that we have a deeper external module indicator (e.g. `import.meta`,
+            // and possibly nested import statements in the future).
+            // Ideally the first few statements will be an import/export anyway.
+            sourceFile.externalModuleIndicator =
+                !(sourceFile.flags & NodeFlags.PossiblyContainsImportMeta) ?
+                    forEach(sourceFile.statements, isAnExternalModuleIndicatorNode) :
+                    walkTreeForExternalModuleIndicators(sourceFile);
+        }
+
+        function isAnExternalModuleIndicatorNode(node: Node) {
+            return hasModifier(node, ModifierFlags.Export)
+                || node.kind === SyntaxKind.ImportEqualsDeclaration && (<ImportEqualsDeclaration>node).moduleReference.kind === SyntaxKind.ExternalModuleReference
+                || node.kind === SyntaxKind.ImportDeclaration
+                || node.kind === SyntaxKind.ExportAssignment
+                || node.kind === SyntaxKind.ExportDeclaration
+                || isMetaProperty(node) && node.keywordToken === SyntaxKind.ImportKeyword && node.name.escapedText === "meta"
                     ? node
-                    : undefined);
+                    : undefined;
+        }
+
+        function walkTreeForExternalModuleIndicators(node: Node): Node {
+            return isAnExternalModuleIndicatorNode(node) ? node : forEachChild(node, walkTreeForExternalModuleIndicators);

```

Comment:
This can accidentally pick up an import/export from an module augmentation as the current file's `externalModuleIndicator` if we're not careful.

```ts
declare module "foo" {
  import * as blah from "blah"
}

import.meta
```